File: day2/day2.rb
#!/usr/bin/ruby

# == Part 1 ==
# Find all lines where `str` does not contain `c` between `X` and `Y` times.
# X-Y c: str

input = File.read('input').split("\n")

input.map! { |l| l.split(/(\d+)-(\d+) (.): (.+)/).drop(1) } # first match w/ regex split is empty
input.map! { |x, y, c, str| [x.to_i, y.to_i, c, str] }

valid_passwords = 0

input.each do |min, max, char, str|
  valid_passwords += 1 if (min..max).include? str.count(char)
end

puts "Valid passwords for range: #{valid_passwords}"

# == Part 2 ==
# Find all lines where `str` contains `c` at either position `X` or `Y` (1-index).
# X-Y c: str

valid_passwords = 0

input.each do |x, y, char, str|
  valid_passwords += 1 if (str[x - 1] == char) ^ (str[y - 1] == char) # ^ is XOR
end

puts "Valid passwords for positions: #{valid_passwords}"

File: day02/day2.c
#include <stdio.h>
#include <stdlib.h>

int str_count(char *str, char target) {
  int count = 0;
  for (int i = 0; str[i]; i++) {
    count += (str[i] == target);
  }
  return count;
}

int main() {
  FILE *infile = fopen("./input", "r");

  int x;
  int y;
  char c;
  char str[30];

  // == Part 1 ==
  // Find all lines where `str` does not contain `c` between `x` and `y` times
  int valid_ranges = 0;

  // == Part 2 ==
  // Find all lines where `str` contains `c` at either position `X` or `Y`
  int valid_positionals = 0;

  while (fscanf(infile, "%d-%d %c: %s", &x, &y, &c, str) != EOF) {
    int count = str_count(str, c);
    if (count >= x && count <= y) {
      valid_ranges++;
    }

    if ((str[x - 1] == c) ^ (str[y - 1] == c)) {
      valid_positionals++;
    }
  }

  printf("Valid passwords for range: %i\n", valid_ranges);
  printf("Valid passwords for positions: %i\n", valid_positionals);

  fclose(infile);
}
